### PR TITLE
Feature/292 refresh azure key vault access token

### PR DESCRIPTION
**Pull request by jekuaitk**
#### Link to ticket

https://leantime.itkdev.dk/#/tickets/showTicket/292

#### Description

Adds check for Azure Key Vault access token expiration. Forces refresh if needed.

### CODE REVIEW

File: src/Service/SF1500Service.php [hunk 1]
```diff
@@ -23,26 +24,39 @@ public function __construct(private readonly CertificateLocator $certificateLoca
...
-            $certificateSettings = $this->options;
+            $this->setupSF1500();
+        }
+
+        if ($this->certificateLocator->shouldAzureKeyVaultAccessTokenBeRefreshed()) {
+            $this->setupSF1500();
+        }
+
+        if (null === $this->sf1500) {
+            throw new SF1500Exception('SF1500 not setup correctly');
```
Comment (rimi-itk) [inline src/Service/SF1500Service.php:35]:
```suggestion
            throw new SF1500Exception('SF1500 not set up correctly');
```

```suggestion
            throw new SF1500Exception('Cannot set up SF1500');
```

File: src/Service/CertificateLocator.php [hunk 2]
```diff
@@ -66,4 +69,20 @@ public function getCertificateLocator(): CertificateLocatorInterface
...
+
+    /**
+     * Checks if certificate locator token should be refreshed.
+     */
+    public function shouldAzureKeyVaultAccessTokenBeRefreshed(): bool
+    {
+        $certificateSettings = $this->options;
+        $locatorType = $certificateSettings['certificate_locator_type'];
+
+        // Reset if token is within a minute of expiration time.
```
Comment (rimi-itk) [inline src/Service/CertificateLocator.php:81]:
```suggestion
        // Refresh if token is within a minute of expiration time.
```

File: src/Service/CertificateLocator.php [hunk 3]
```diff
@@ -66,4 +69,20 @@ public function getCertificateLocator(): CertificateLocatorInterface
...
+        if (self::LOCATOR_TYPE_AZURE_KEY_VAULT === $locatorType && $this->tokenExpiration - 60 <= (new \DateTimeImmutable())->getTimestamp()) {
+            return true;
+        }
+
+        return false;
```
Comment (rimi-itk) [inline src/Service/CertificateLocator.php:86]:
```suggestion
        return self::LOCATOR_TYPE_AZURE_KEY_VAULT === $locatorType
          && $this->tokenExpiration - 60 <= (new \DateTimeImmutable())->getTimestamp();
```

File: src/Service/SF1500Service.php [hunk 4]
```diff
@@ -23,26 +24,39 @@ public function __construct(private readonly CertificateLocator $certificateLoca
     public function getSF1500(): SF1500
     {
         if (null === $this->sf1500) {
-            $certificateSettings = $this->options;
+            $this->setupSF1500();
+        }
+
+        if ($this->certificateLocator->shouldAzureKeyVaultAccessTokenBeRefreshed()) {
+            $this->setupSF1500();
+        }
```
Comment (rimi-itk) [inline src/Service/SF1500Service.php:32]: Consider merging this with the previous `if` statement, i.e.

```php
        if (null === $this->sf1500
            || $this->certificateLocator->shouldAzureKeyVaultAccessTokenBeRefreshed()) {
            $this->setupSF1500();
        }
```

File: src/Service/CertificateLocator.php [hunk 5]
```diff
@@ -66,4 +69,20 @@ public function getCertificateLocator(): CertificateLocatorInterface
 
         throw new CertificateLocatorException(sprintf('Invalid certificate locator type: %s', $locatorType));
     }
+
+    /**
+     * Checks if certificate locator token should be refreshed.
+     */
+    public function shouldAzureKeyVaultAccessTokenBeRefreshed(): bool
```
Comment (rimi-itk) [inline src/Service/CertificateLocator.php:76]: Consider renaming this function to `isExpired` or `needsRefresh`, i.e. don't mention “Azure Key Vault”. Other (theoretical/future) locators may need refreshment as well.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: src/Service/SF1500Service.php (hunk 1 above)
Comment (jekuaitk) [reply src/Service/SF1500Service.php:35]: Fixed.

File: src/Service/CertificateLocator.php (hunk 2 above)
Comment (jekuaitk) [reply src/Service/CertificateLocator.php:81]: Fixed.

File: src/Service/CertificateLocator.php (hunk 3 above)
Comment (jekuaitk) [reply src/Service/CertificateLocator.php:86]: Fixed.

File: src/Service/CertificateLocator.php (hunk 5 above)
Comment (jekuaitk) [reply src/Service/CertificateLocator.php:76]: Renamed to `needsRefresh`.

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**